### DISCUSSION TITLE
avformat input_queue_size

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#2994] [Q&A]
When using `video_source avformat,android_camera,1` I sometimes get to log:
```
2024-04-14 17:14:28.136 18268-18321 Baresip+ Lib            com.tutpro.baresip.plus              D  Input queue was full, dropping frame, consider raising the input_queue_size option (current value: 5)
```
I then checked `ffmpeg/libavdevice/android_camera.c` and found the option:
```
static const AVOption options[] = {
    { "video_size", "set video size given as a string such as 640x480 or hd720", OFFSET(requested_width), AV_OPT_TYPE_IMAGE_SIZE, {.str = NULL}, 0, 0, DEC },
    { "framerate", "set video frame rate", OFFSET(framerate), AV_OPT_TYPE_VIDEO_RATE, {.str = "30"}, 0, INT_MAX, DEC },
    { "camera_index", "set index of camera to use", OFFSET(camera_index), AV_OPT_TYPE_INT, {.i64 = 0}, 0, INT_MAX, DEC },
    { "input_queue_size", "set maximum number of frames to buffer", OFFSET(input_queue_size), AV_OPT_TYPE_INT, {.i64 = 5}, 0, INT_MAX, DEC },
    { NULL },
};
```
However, `modules/avformat/avformat.c`  function
```
avformat_shared_alloc(struct shared **shp, const char *dev,
			  double fps, const struct vidsz *size,
			  bool video)
```
does not have the input queue size parameter.

One way to set that parameter might be by adding it video_source parameter:
```
video_source avformat,android_camera[,<input_queue_size>]
```
Currently the parsing of the parameters goes like this:
```
if (0 == re_regex(dev, str_len(dev), "[^,]+,[^]+", &pl_fmt, &pl_dev)) 
```
Any hints on how the change the regex so that optional input queue size value could be assigned to new &pl_iqs argument? 

Or what would a better option to allow specifying the input_queue_size?

**Comment by sreimers** [comment]
No experience with this specific AVOption but how I understand `1/30 Frames = 33ms` and  `5 Frames * 33ms = 165ms` buffer Latency, so increasing `input_queue_size` would lead to much higher latencies. I would guess that encoding is not fast enough, so reducing the framerate or other speed improvements (like reducing codec quality) could help.

**Reply by juha-h** [reply]
Helped when I used smaller video_size.
